DRAFT: Task/7 5 2025 fmt string followup

emily-howell opened the pull request:
Looking to address some of the issues uncovered with the original implementation. 

Reverts #1511 and actual implementation for extension checking is added here

#1522 and #1518 document issues with the default value using family and not cycle which is also addressed here.

DRAFT subject to changes and discussion on requirements for path formatting

emily-howell committed:
Fix conflicting family value found in output test file names

emily-howell committed:
Reverting my temporary change to the actions dump file name. Ideal is for fmt to take into account extension.

emily-howell committed:
Adding handling for specifying extensions

emily-howell committed:
Dissiminating the extension check

emily-howell committed:
Modifying so that when no format provided it will append the cycle rather than the family value

emily-howell committed:
Also making default format value use cycle variable instead of family

cyrush commented on src/libs/ascent/runtimes/flow_filters/ascent_runtime_rover_filters.cpp:
```diff
@@ -335,11 +345,11 @@ RoverXRay::execute()
       if(cycle != -1)
       {
-        tracer.save_bov(expand_path_special_variables(bov_filename, mpi_comm_id, cycle));
+        tracer.save_bov(expand_path_special_variables(bov_filename, ".png", mpi_comm_id, cycle));
```
this case would be `".bov"`

emily-howell committed:
Decoupling cycle and family values

emily-howell committed:
Now supports cases where no formatting is given

emily-howell committed:
Adding a test to make sure things work when only a format type is given

emily-howell committed:
family value will work both when extensions added or not for a specific format

emily-howell committed:
Fixing bov file extension

emily-howell committed:
reverting accidental test file add

emily-howell commented:
The current state of this PR includes the following changes:
- [x] Reverts the patch fix made in #1511 
- [x] Tracks the file path extensions in family value checks
  - [x] This also fixes the failing tests caused by file conflicts
- [x] Default value is cycle again rather than family value
- [x] Family value has been decoupled from cycle. Can now pass a family value seed as an option that gets stored in the metadata
- [x] Patterns not including a format such as "filename{cycle}" and "filename{cycle:}" are now supported
- [x] Fixed extension for bov files 

Outstanding:
- [x] Using the family value as a backup if cycle not available for default formatting. Concern that this might be confusing to users still stands.
- [x] Underscore added to default
- [x] default should follow format set by table
- [ ] Improve documentation
- [ ] Add a set of tests that does everything in the "table" and documentation

emily-howell committed:
Adding a test that tests the family_value_seed_option

cyrush commented on src/libs/ascent/utils/ascent_string_utils.cpp:
```diff
@@ -50,6 +50,28 @@ void split_string(const std::string &s,
+  std::vector<std::string> elems;
+  detail::split_string(s, delim, elems);
+  return elems;
+}
```
You can use Conduit's string splitter util

https://github.com/LLNL/conduit/blob/b8f4ad0858980e430661b2d381645bd1de5c0e4b/src/libs/conduit/conduit_utils.hpp#L350

emily-howell committed:
Merge branch 'develop' of github.com:Alpine-DAV/ascent into task/7_5_2025_fmt_string_followup

emily-howell committed:
Addressing a number of requested changes from PR including:
1) Adding an underscore before the default value if one was not already added
2) Setting default formats for each data type (time is g, cycle is 06d, and family is 06d)
3) Using family as a backup if cycle is not given

These changes will break a number of existing tests so a followup commit will fix other tests

emily-howell committed:
Updating testing image comparison and removal functions to reflect path changes

emily-howell committed:
Updating baseline image paths

emily-howell committed:
Using an overloading approach instead so underscore is only added when default number used for image test paths

emily-howell committed:
Fixing a number of tests that had cycle 0 rather than 100

emily-howell committed:
Updating the ASCENT_ACTIONS_DUMP... which I thought had already been done...

emily-howell committed:
Merge branch 'develop' of github.com:Alpine-DAV/ascent into task/7_5_2025_fmt_string_followup

emily-howell committed:
Fixing new rover tests to use updated path format

emily-howell committed:
Fixing the hola mpi test

emily-howell committed:
Making it so we can pass a cycle number to the check_test_image and remove_test_image functions

emily-howell committed:
Fixing missnamed file

emily-howell commented:
@cyrush I think this has all the requests (minus robust documentation which you said should go in a future PR). I added the family value as a fallback, default formats (`06d` for cycle and family `g` for time) and an `_` in front of the default fmt value iff the path does not already end with a `_` character.

As a note, those changes, both the default format of `06d` and the `_`, changed how the vast majority of our output images from tests are named so there are a lot of renamed baseline_images but none of the images themselves have been changed.

I also got annoyed with how `check_test_image` and `remove_test_image` worked so I added an overload option for both that can take a cycle number in as an int and format it. This is nice because it means that we can make changes to the formatting and only have to change that one function rather than every instance of the function call. I did notice that a number of places that use a custom value for `check_test_image` don't also pass that value to `remove_test_image` so that should get fixed. For the `ASCENT_ACTIONS_DUMP` I don't think you can overload macros? so I made a few different options that also allow you to specify the cycle and format.

emily-howell committed:
Updating the actions dump so it can also take in a number and format

cyrush commented on src/tests/t_utils.hpp:
```diff
@@ -682,16 +705,25 @@ add_matset_to_spiral(Node &n_mesh, const int ndomains)
+
+// Macro to save ascent actions file with default cycle and formatting
+#define ASCENT_ACTIONS_DUMP(actions,name,msg) \
+  ASCENT_ACTIONS_DUMP_CYCLE(actions,name,msg,100)
```
these macros are nice

cyrush reviewed (approved):
Looks good, thanks for working through this!

emily-howell committed:
Merge branch 'develop' of github.com:Alpine-DAV/ascent into task/7_5_2025_fmt_string_followup

emily-howell committed:
Changes to get merged rover changes working with branch updates

emily-howell committed:
Fixing other tests to match this branch's updates

emily-howell committed:
Adding back reduced tolerance for rover that I accidentally removed  despite the explicit comment explaining why it is there :)

emily-howell committed:
Merge branch 'develop' of github.com:Alpine-DAV/ascent into task/7_5_2025_fmt_string_followup
